pulsar-broker/src/main/java/org/apache/pulsar/broker/s3offload/S3ManagedLedgerOffloader.java

```diff
@@ -60,28 +80,119 @@ public static S3ManagedLedgerOffloader create(ServiceConfiguration conf,
+            try {
+                long startEntry = 0;
+                int partId = 1;
+                long entryBytesWritten = 0;
+                List<PartETag> etags = new LinkedList<>();
+                while (startEntry <= readHandle.getLastAddConfirmed()) {
+                    int blockSize = BlockAwareSegmentInputStreamImpl
+                        .calculateBlockSize(maxBlockSize, readHandle, startEntry, entryBytesWritten);
+
+                    try (BlockAwareSegmentInputStream blockStream = new BlockAwareSegmentInputStreamImpl(
+                        readHandle, startEntry, blockSize)) {
+
+                        UploadPartResult uploadRes = s3client.uploadPart(
+                            new UploadPartRequest()
+                                .withBucketName(bucket)
+                                .withKey(dataBlockKey)
+                                .withUploadId(dataBlockRes.getUploadId())
+                                .withInputStream(blockStream)
+                                .withPartSize(blockSize)
+                                .withPartNumber(partId));
+                        etags.add(uploadRes.getPartETag());
+                        indexBuilder.addBlock(startEntry, partId, blockSize);
+
+                        if (blockStream.getEndEntryId() != -1) {
+                            startEntry = blockStream.getEndEntryId() + 1;
+                        } else {
+                            // could not read entry from ledger.
+                            break;
+                        }
+                        entryBytesWritten += blockStream.getBlockEntryBytesCount();
+                        partId++;
+                    }
+                }
+
+                s3client.completeMultipartUpload(new CompleteMultipartUploadRequest()
+                    .withBucketName(bucket).withKey(dataBlockKey)
+                    .withUploadId(dataBlockRes.getUploadId())
+                    .withPartETags(etags));
+            } catch (Throwable t) {
+                s3client.abortMultipartUpload(
+                    new AbortMultipartUploadRequest(bucket, dataBlockKey, dataBlockRes.getUploadId()));
+                promise.completeExceptionally(t);
+                return;
+            }
+
+            // upload index block
+            try (OffloadIndexBlock index = indexBuilder.build();
+                 InputStream indexStream = index.toStream()) {
+                // write the index block
+                ObjectMetadata metadata = new ObjectMetadata();
+                metadata.setContentLength(indexStream.available());
+                s3client.putObject(new PutObjectRequest(
+                    bucket,
+                    indexBlockOffloadKey(readHandle, uuid),
+                    indexStream,
+                    metadata));
+                promise.complete(null);
+            } catch (Throwable t) {
+                if (s3client.doesObjectExist(bucket, dataBlockOffloadKey(readHandle, uuid))) {
```

Comment:
there's no need to check. in S3 deleting an object that doesn't exist will return success.
https://docs.aws.amazon.com/AmazonS3/latest/API/multiobjectdeleteapi.html

Use deleteObjects() so that only one call is made.

Reply:
Thanks, this is to avoid findify.s3mock error as above.